### PR TITLE
Trade

### PR DESCRIPTION
**Pull request by snreynolds**

### CODE REVIEW

File: src/entities/route.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        this.protocol = Protocol.V2
+
+    }
+
+}
+
+//v3 route wrapper
+export class RouteV3<TInput extends Currency, TOutput extends Currency> extends V3RouteSDK<TInput, TOutput> implements IRoute<TInput, TOutput> {
+    public readonly protocol: Protocol
+    public readonly pools: Pool[]
```
Comment (willpote) [inline src/entities/route.ts:33]: Not needed since its provided by V3RouteSDK

File: src/entities/route.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    constructor(v2Route: V2RouteSDK<TInput, TOutput>) {
+        super(v2Route.pairs, v2Route.input, v2Route.output)       
+        this.pools = this.pairs
+        this.protocol = Protocol.V2
+
+    }
+
+}
+
+//v3 route wrapper
```
Comment (willpote) [inline src/entities/route.ts:30]: Nit: `// V3`

File: src/entities/trade.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        routev2,
+        inputAmount,
+        outputAmount,
+      })
+    }
+
+    for (const { routev3, amount } of v3Routes) {
+      const v3Trade = V3TradeSDK.fromRoute(routev3, amount, tradeType)
+      const inputAmount = (await v3Trade).inputAmount
+      const outputAmount = (await v3Trade).outputAmount
```
Comment (willpote) [inline src/entities/trade.ts:223]: Nit:
```
const v3Trade = await V3TradeSDK.fromRoute(...);
const { inputAmount, outputAmount } = v3Trade;
```

can do similar in the v2 loop too

File: src/entities/trade.ts [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    if (route instanceof V2RouteSDK) {
+      const v2Trade = new V2TradeSDK(route, amount, tradeType)
+      const inputAmount = v2Trade.inputAmount
+      const outputAmount = v2Trade.outputAmount
+      v2Routes = [{ routev2: route, inputAmount, outputAmount }]
+      v3Routes = []
+    } else {
+      const v3Trade = V3TradeSDK.fromRoute(route, amount, tradeType)
+      const inputAmount = (await v3Trade).inputAmount
+      const outputAmount = (await v3Trade).outputAmount
```
Comment (willpote) [inline src/entities/trade.ts:265]: Same

**Review by willpote — COMMENTED** [review]
Looks good (minus tests)

There some odd spacing in `route.ts` btw, could you run prettier on the PR?

Also might be worth checking with @justindomingue that the Trade object has everything the interface will need (if you haven't already)

File: src/entities/route.ts [hunk 5]
```diff
@@ -1,44 +1,42 @@
...
 export interface IRoute<TInput extends Currency, TOutput extends Currency> {
-    protocol: Protocol
-    // array of pools if v3 or pairs if v2
-    pools : Pool[] | Pair[]
-    path : Token[]
-    midPrice : Price<TInput, TOutput>
-
+  protocol: Protocol
+  // array of pools if v3 or pairs if v2
+  pools: Pool[] | Pair[]
```
Comment (justindomingue) [inline src/entities/route.ts:9]: I wonder if you could make that a generic?

```
export interface IRoute<TInput extends Currency, TOutput extends Currency, TPoolItem> {
  pools: TPoolItem[]
```

**Pull request merged**